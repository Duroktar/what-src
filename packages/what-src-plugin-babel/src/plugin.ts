import template from '@babel/template'
import { isNullOrUndefined, getIn } from './utils'
import { defaultOptions } from './options'
import * as T from './types'

let nextId = 1
let disabled = false

export const babelPlugin = ({ types: t }: T.BabelPluginContext): T.BabelPlugin => ({
  pre(): void {
    const opts: Required<T.WhatSrcPluginOptions> = Object.assign({}, defaultOptions, this.opts)

    if (!disabled && process.env.NODE_ENV === 'production' && !opts.productionMode) {
      console.log(
        '@what-src/babel-plugin - running in production mode is disabled. ' +
        'To enable set the \'productionMode\' configuration option to true.'
      )
      disabled = true
    };

    this.cache = {}
  },
  post(state): void {
    if (disabled) return

    const opts: Required<T.WhatSrcPluginOptions> = Object.assign({}, defaultOptions, this.opts)

    const eventListener = template.statement(`
      try {
        const cache = JSON.parse(%%cache%%);
        window[%%globalCacheKey%%] = function (e) {
          if (e.metaKey) {
            if (%%stopPropagation%%) e.stopPropagation()
            if (%%preventDefault%%) e.preventDefault()
            const xhr = new XMLHttpRequest();
            xhr.open('POST', %%serverUrl%%, true);
            xhr.setRequestHeader('Content-type', 'application/json');
            const dataset = cache[e.path[0].dataset['WhatSrc']]
            if (typeof dataset !== 'undefined') xhr.send(dataset);
          }
        }
        window.document.removeEventListener('click', window[%%globalCacheKey%%])
        window.document.addEventListener('click', window[%%globalCacheKey%%])
      } catch {}
    `)

    const rawCache = JSON.stringify(this.cache)

    const ast = eventListener({
      globalCacheKey: t.stringLiteral(opts.globalCacheKey),
      serverUrl: t.stringLiteral(opts.serverUrl),
      cache: t.stringLiteral(rawCache),
      stopPropagation: t.booleanLiteral(opts.stopPropagation),
      preventDefault: t.booleanLiteral(opts.preventDefault),
    })

    state.path.node.body.push(ast)
  },
  visitor: {
    JSXElement: {
      enter(path, state): void {
        const isFragment = getIn('name.property.name', path.node.openingElement) === 'Fragment'
        if (disabled || isFragment || isNullOrUndefined(path.node.openingElement.loc)) return

        const opts: Required<T.WhatSrcPluginOptions> = Object.assign(
          {}, defaultOptions, state.opts)
        const meta = path.node.openingElement.loc.start

        const metaData = JSON.stringify({
          filename: state.filename,
          line:     meta.line,
          column:   meta.column + 1, // don't know why we need the '+ 1' but we do *le shrug
        })

        const attr = t.jsxAttribute(
          t.jsxIdentifier(opts.dataTag),
          t.stringLiteral(nextId.toString()),
        )

        path.node.openingElement.attributes.push(attr)

        this.cache[nextId] = metaData

        nextId++
      },
    },
  },
})
