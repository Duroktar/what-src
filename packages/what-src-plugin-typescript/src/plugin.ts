import ts from "typescript";
import path from "path";
import * as H from "@what-src/plugin-core";

type UserOptions = { importName: string; importFrom: string };

export function createTransformer(opts: UserOptions = {} as any) {
  let basedir = '/Users/duroktar/fun/what-src-webpack-plugin/packages/what-src-example-typescript-loader/src/'
  const options = H.getAllPluginOptions({});
  const resolver = H.getResolver({ options, basedir, cache: {} });
  const cacheFile = opts.importFrom || 'what-src-cache'
  const importName = opts.importName || '__WhatSrcGlobalVariable'

  const transformer: ts.TransformerFactory<ts.SourceFile> = context => {
    const { module, outDir } = context.getCompilerOptions()
    const cacheFilePath = path.join(outDir || '', cacheFile)

    // !!! Each file
    return sf => {

      const visitor: ts.Visitor = node => {

        if (ts.isSourceFile(node)) {
          // <3 <3 <3 https://github.com/Microsoft/TypeScript/issues/18369#issuecomment-330133796

          // const WhatSrcGlobalVariable = require('../../what-src-cache.js')   <-- example
          if (module === ts.ModuleKind.CommonJS) {
            const file = (node as ts.Node) as ts.SourceFile
            node = ts.updateSourceFileNode(file, [
              ts.createVariableStatement(
                /*modifiers*/ undefined,
                ts.createVariableDeclarationList([
                  ts.createVariableDeclaration(
                    importName,
                    /*type*/ undefined,
                    ts.createPropertyAccess(
                      ts.createCall(
                        ts.createIdentifier('require'),
                        [],
                        [ts.createLiteral(cacheFilePath)]
                      ),
                      ts.createIdentifier('default')
                    )
                  ),
                ])
              ),
              ...file.statements
            ])
          }

          // import WhatSrcGlobalVariable from '../../what-src-cache.js'   <-- example
          if (module === ts.ModuleKind.ES2015 || module === ts.ModuleKind.ESNext) {
            const file = (node as ts.Node) as ts.SourceFile
            node = ts.updateSourceFileNode(file, [
              ts.createImportDeclaration(
                /* decorators */ undefined,
                /* modifiers */ undefined,
                ts.createImportClause(ts.createIdentifier(importName), undefined),
                ts.createLiteral(cacheFilePath)
              ),
              ...file.statements
            ])
          }
        }

        if (ts.isJsxElement(node) && !ts.isJsxFragment(node)) {
          if (ts.isJsxOpeningElement(node.openingElement)) {
            const { character: col, line } = sf.getLineAndCharacterOfPosition(
              node.openingElement.getStart()
            );

            const nextId = resolver.resolve({ col, line: line + 1, basedir }, sf.fileName);

            const attrs = ts.updateJsxAttributes(node.openingElement.attributes, [
              ts.createJsxAttribute(
                ts.createIdentifier(options.dataTag),
                ts.createStringLiteral(nextId as any)
              )])
              node.openingElement.attributes = attrs
          }
        }

        ts.forEachChild(node, n => {
          if (!n.parent) n.parent = node;
        });

        return ts.visitEachChild(node, visitor, context);
      };

      const last = ts.visitNode(sf, visitor);

      resolver.emit(cacheFilePath)

      return last
    };
  };

  return transformer;
}

export default createTransformer;
