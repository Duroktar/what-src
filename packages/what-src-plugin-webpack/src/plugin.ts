import * as path from 'path'
import * as process from 'process'
import * as childProcess from 'child_process'
import * as webpack from 'webpack'
import chalk from 'chalk'

import {
  getWhatSrcWebpackPluginHooks,
  legacyHookMap,
  WhatSrcHooks,
} from './hooks'
import { createLogger } from './logger'
import { Options, Logger } from './types'
import { defaultOptions } from './options'

const PLUGIN_NAME = 'what-src-server-plugin'

class WhatSrcServerWebpackPlugin {
  public static getCompilerHooks(
    compiler: any
  ): Record<WhatSrcHooks, any> {
    return getWhatSrcWebpackPluginHooks(compiler)
  }

  public readonly options: Options

  private compiler: any = undefined;
  private doneCallback: () => void = () => {};
  private isRunning: boolean = false;
  private logger: Logger;
  private service?: childProcess.ChildProcess;

  constructor(options?: Partial<Options>) {
    options = options || ({} as Options)
    this.options = { ...defaultOptions, ...options }

    this.logger = options.logger || createLogger()

    if (this.showProductionWarning) {
      this.logger.info(
        '\n@what-src/webpack-plugin - running in production mode is disabled. ' +
        'To enable set the \'productionMode\' configuration option to true.'
      )
    };
  }

  public apply(compiler: any) {
    if (process.env.NODE_ENV === 'production' && !this.options.productionMode) {
      return
    }

    this.compiler = compiler

    // validate logger
    if (this.logger) {
      if (!this.logger.error || !this.logger.warn || !this.logger.info) {
        throw new Error(
          "Invalid logger object - doesn't provide `error`, `warn` or `info` method."
        )
      }
    }

    this.pluginStart()
    this.pluginStop()
    this.pluginCompile()
    this.pluginEmit()
    this.pluginDone()
  }

  private pluginStart() {
    const run = (_compiler: webpack.Compiler, callback: () => void) => {
      this.isRunning = false
      callback()
    }

    const watchRun = (_compiler: webpack.Compiler, callback: () => void) => {
      this.isRunning = true
      callback()
    }

    if ('hooks' in this.compiler) {
      // webpack 4+
      this.compiler.hooks.run.tapAsync(PLUGIN_NAME, run)
      this.compiler.hooks.watchRun.tapAsync(PLUGIN_NAME, watchRun)
    } else {
      // webpack 2 / 3
      this.compiler.plugin('run', run)
      this.compiler.plugin('watch-run', watchRun)
    }
  }

  private pluginStop() {
    const watchClose = () => {
      this.killService()
    }

    const done = (_stats: webpack.Stats) => {
      if (!this.isRunning) {
        this.killService()
      }
    }

    if ('hooks' in this.compiler) {
      // webpack 4+
      this.compiler.hooks.watchClose.tap(PLUGIN_NAME, watchClose)
      this.compiler.hooks.done.tap(PLUGIN_NAME, done)
    } else {
      // webpack 2 / 3
      this.compiler.plugin('watch-close', watchClose)
      this.compiler.plugin('done', done)
    }
  }

  private pluginCompile() {
    if ('hooks' in this.compiler) {
      // webpack 4+
      const WhatSrcHooks = WhatSrcServerWebpackPlugin.getCompilerHooks(
        this.compiler
      )
      this.compiler.hooks.compile.tap(PLUGIN_NAME, () => {
        WhatSrcHooks.serviceBeforeStart.callAsync(() => {
          if (!this.service || !this.service.connected) {
            this.spawnService()
          }
        })
      })
    } else {
      // webpack 2 / 3
      this.compiler.plugin('compile', () => {
        this.compiler.applyPluginsAsync(
          legacyHookMap.serviceBeforeStart,
          () => {
            if (!this.service || !this.service.connected) {
              this.spawnService()
            }
          }
        )
      })
    }
  }

  private pluginEmit() {
    const emit = (compilation: any, callback: () => void) => {
      callback()
    }

    if ('hooks' in this.compiler) {
      // webpack 4+
      this.compiler.hooks.emit.tapAsync(PLUGIN_NAME, emit)
    } else {
      // webpack 2 / 3
      this.compiler.plugin('emit', emit)
    }
  }

  private pluginDone() {
    if ('hooks' in this.compiler) {
      // webpack 4+
      this.compiler.hooks.done.tap(
        PLUGIN_NAME,
        (_stats: webpack.Stats) => {
          this.doneCallback()
        }
      )
    } else {
      // webpack 2 / 3
      this.compiler.plugin('done', () => {
        this.doneCallback()
      })
    }
  }

  private spawnService() {
    const env: { [key: string]: string | undefined } = {
      ...process.env,
      CONTEXT: this.compiler.options.context,
      WHAT_SRC_DAEMON_HOST: JSON.stringify(this.options.host),
      WHAT_SRC_DAEMON_PORT: JSON.stringify(this.options.port),
      WHAT_SRC_DAEMON_ENDPOINT: JSON.stringify(this.options.endpoint),
      WHAT_SRC_DAEMON_EDITOR: JSON.stringify(this.options.editor),
      WHAT_SRC_DAEMON_SHH: JSON.stringify(this.options.shh),
    }

    this.service = childProcess.fork(
      path.resolve(
        __dirname,
        './service.js'
      ),
      [],
      {
        env,
        execArgv: [].concat(/* this.nodeArgs */[]),
        stdio: ['inherit', 'inherit', 'inherit', 'ipc'],
      }
    )

    if ('hooks' in this.compiler) {
      // webpack 4+
      const WhatSrcHooks = WhatSrcServerWebpackPlugin.getCompilerHooks(
        this.compiler
      )
      WhatSrcHooks.serviceStart.call(
        'serviceStart',
      )
    } else {
      // webpack 2 / 3
      this.compiler.applyPlugins(
        legacyHookMap.serviceStart,
        'serviceStart',
      )
    }

    if (this.logger && !this.options.shh) {
      this.logger.info(
        'server is running at ' +
        chalk.cyanBright.bold(
          this.options.host + ':' +
            this.options.port + '/' +
            this.options.endpoint
        )
      )
    }

    this.service.on('exit', (code: string | number, signal: string) =>
      this.handleServiceExit(code, signal)
    )
  }

  private killService() {
    if (!this.service) {
      return
    }
    try {
      this.service.kill()
      this.service = undefined
    } catch (e) {
      if (this.logger && !this.options.shh) {
        this.logger.error(e)
      }
    }
  }

  private handleServiceExit(_code: string | number, signal: string) {
    if (signal !== 'SIGABRT') {
      return
    }
    // probably out of memory :/
    if (this.logger && !this.options.shh) {
      this.logger.error(
        chalk.red(
          '@what-src/express aborted - probably out of memory..'
        )
      )
    }
  }

  private get mode() {
    return process.env.NODE_ENV
  }

  private get showProductionWarning() {
    const { productionMode, shh } = this.options
    return (this.mode === 'production' && !productionMode) && !shh
  }
}

export { WhatSrcServerWebpackPlugin }
