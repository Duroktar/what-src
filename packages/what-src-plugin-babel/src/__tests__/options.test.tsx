import { defaultOptions } from '@what-src/plugin-core'
import { WhatSrcBabelPlugin } from '../plugin'

const plugin = new WhatSrcBabelPlugin()
const dataTag = plugin.options.dataTag
const testTag = 'TEST_DATA_TAG'

describe('what-src babel-plugin configuration', () => {
  it('is stable', () => {
    expect(plugin.options).toMatchSnapshot()
  })

  it('has defaults', () => {
    expect(plugin.options).toEqual(defaultOptions)
  })

  it('is initially configurable', () => {
    const plugin2 = new WhatSrcBabelPlugin({ dataTag: testTag })
    expect(plugin2.options.dataTag).toEqual(testTag)
    plugin2.options.dataTag = dataTag
    expect(plugin2.options.dataTag).toEqual(dataTag)
  })

  it('is later configurable', () => {
    plugin.options.dataTag = testTag
    expect(plugin.options.dataTag).toEqual(testTag)
    plugin.options.dataTag = dataTag
    expect(plugin.options.dataTag).toEqual(dataTag)
  })
})
