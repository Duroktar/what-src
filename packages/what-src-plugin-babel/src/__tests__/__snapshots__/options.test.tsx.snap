// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`what-src babel-plugin configuration is stable 1`] = `
Object {
  "blacklistedTags": Array [
    "Helmet",
    "head",
    "meta",
    "script",
  ],
  "dataTag": "data-what-src",
  "enableXkcdMode": false,
  "globalCacheKey": "__what-src-global-callback-key",
  "importFrom": "what-src-cache.jss",
  "importName": "__WhatSrcGlobalVariable",
  "preventDefault": true,
  "productionMode": false,
  "serverUrl": "http://localhost:8018/__what_src",
  "stopPropagation": true,
  "useRemote": false,
}
`;
